### PR TITLE
Provoking vertex validation

### PR DESCRIPTION
**Pull request by locke-lunarg**
closes: https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/2755

Added two validations.
```
VUID-VkPipelineRasterizationProvokingVertexStateCreateInfoEXT-provokingVertexMode-04883
If provokingVertexMode is VK_PROVOKING_VERTEX_MODE_LAST_VERTEX_EXT, then the provokingVertexLast feature must be enabled
```
```
VUID-vkCmdBindPipeline-pipelineBindPoint-04881
If pipelineBindPoint is VK_PIPELINE_BIND_POINT_GRAPHICS and the provokingVertexModePerPipeline limit is VK_FALSE, then pipeline’s VkPipelineRasterizationProvokingVertexStateCreateInfoEXT::provokingVertexMode must be the same as that of any other pipelines previously bound to this bind point within the current renderpass instance, including any pipeline already bound when beginning the renderpass instance
```
 The extension `VK_EXT_PROVOKING_VERTEX_EXTENSION_NAME` is new. The GPU driver doesn't support it, so the test only can run in mock_icd.

### CODE REVIEW

File: layers/core_validation.cpp [hunk 1]
```diff
@@ -1681,6 +1681,16 @@ bool CoreChecks::ValidatePipelineUnlocked(const PIPELINE_STATE *pPipeline, uint3
                 }
             }
         }
+
+        auto provoking_vertex_state_ci = lvl_find_in_chain<VkPipelineRasterizationProvokingVertexStateCreateInfoEXT>(
```
Comment (ncesario-lunarg) [inline layers/core_validation.cpp:1685]: I think these got refactored to `LvlFindInChain` to better conform with the coding guidelines. I think MarkL left these in for backwards compatibility.

File: layers/state_tracker.cpp [hunk 2]
```diff
@@ -2146,13 +2156,16 @@ void ValidationStateTracker::PostCallRecordCreateDevice(VkPhysicalDevice gpu, co
     if (!state_tracker->device_extensions.vk_feature_version_1_2 && state_tracker->api_version >= VK_API_VERSION_1_1) {
         // Get the needed subgroup limits
         auto subgroup_prop = LvlInitStruct<VkPhysicalDeviceSubgroupProperties>();
+        auto provoking_vertex_prop = LvlInitStruct<VkPhysicalDeviceProvokingVertexPropertiesEXT>();
+        subgroup_prop.pNext = &provoking_vertex_prop;
```
Comment (ncesario-lunarg) [inline layers/state_tracker.cpp:2160]: `auto subgroup_prop = LvlInitStruct<VkPhysicalDeviceSubgroupProperties>(&provoking_vertex_prop);` if you want to.

File: tests/vklayertests_pipeline_shader.cpp [hunk 3]
```diff
@@ -10222,3 +10222,100 @@ TEST_F(VkLayerTest, ComputePipelineInvalidFlags) {
...
+    } else {
+        printf("%s Extension %s is not supported.\n", kSkipPrefix, VK_EXT_PROVOKING_VERTEX_EXTENSION_NAME);
+        return;
+    }
+
+    auto provoking_vertex_properties = lvl_init_struct<VkPhysicalDeviceProvokingVertexPropertiesEXT>();
+    auto properties2 = lvl_init_struct<VkPhysicalDeviceProperties2>(&provoking_vertex_properties);
+    vk::GetPhysicalDeviceProperties2(gpu(), &properties2);
+
+    if (provoking_vertex_properties.provokingVertexModePerPipeline == VK_TRUE) {
```
Comment (ncesario-lunarg) [inline tests/vklayertests_pipeline_shader.cpp:10271]: Can you not set this to `VK_FALSE` to explicitly disable the feature?

**Review by ncesario-lunarg — APPROVED** [review]
Some minor nits, but LGTM!

File: layers/core_validation.cpp (hunk 1 above)
Comment (locke-lunarg) [reply layers/core_validation.cpp:1685]: fixed

File: layers/state_tracker.cpp (hunk 2 above)
Comment (locke-lunarg) [reply layers/state_tracker.cpp]: fixed

File: tests/vklayertests_pipeline_shader.cpp (hunk 3 above)
Comment (locke-lunarg) [reply tests/vklayertests_pipeline_shader.cpp]: fixed

File: tests/vklayertests_pipeline_shader.cpp [hunk 4]
```diff
@@ -10222,3 +10222,96 @@ TEST_F(VkLayerTest, ComputePipelineInvalidFlags) {
...
+
+    if (DeviceExtensionSupported(gpu(), nullptr, VK_EXT_PROVOKING_VERTEX_EXTENSION_NAME)) {
+        m_device_extension_names.push_back(VK_EXT_PROVOKING_VERTEX_EXTENSION_NAME);
+    } else {
+        printf("%s Extension %s is not supported.\n", kSkipPrefix, VK_EXT_PROVOKING_VERTEX_EXTENSION_NAME);
+        return;
+    }
+
+    auto provoking_vertex_properties = LvlInitStruct<VkPhysicalDeviceProvokingVertexPropertiesEXT>();
+    provoking_vertex_properties.provokingVertexModePerPipeline = VK_FALSE;
```
Comment (ncesario-lunarg) [inline tests/vklayertests_pipeline_shader.cpp:10268]: I think you may want to set this _after_ `GetPhysicalDeviceProperties2` (as I think that will set the value?), but before InitState/CreateDevice?

**Review by ncesario-lunarg — APPROVED** [review]
One last nit in the test, but LGTM!

File: tests/vklayertests_pipeline_shader.cpp (hunk 4 above)
Comment (locke-lunarg) [reply tests/vklayertests_pipeline_shader.cpp]: Changed to the pervious version.

**Review by Tony-LunarG — APPROVED** [review]
LGTM

**Pull request merged**